Reduce heap allocations in `auparse.go`

inge4pres opened the pull request:
## Reason for this PR
We noticed high GC pressure (recorded by profiling) when running `auditbeat` and an eBPF-powered profiling agent.
When auditbeat is configured to read _all_ BPF syscalls, and a profiling agent based on eBPF executes many, a relevant CPU consumption goes into parsing and sending the increased number of events from auditd.

## Details
With this PR we have a reduction of ~15% of heap allocations in the parse/enrich phases of `AuditMessage`.
The commits contain some additional benchmarks and experiments that lead to the improvement.

The benchmark output
```
# before
BenchmarkExtractKeyValuePairs/bpf_message_current-8         	  113053	      9687 ns/op	    2198 B/op	      33 allocs/op
BenchmarkExtractKeyValuePairs/short_message_current-8       	  292078	      3440 ns/op	    1232 B/op	       9 allocs/op

# after
BenchmarkExtractKeyValuePairs/bpf_message_current-8         	  118813	      8945 ns/op	    2616 B/op	      24 allocs/op
BenchmarkExtractKeyValuePairs/short_message_current-8       	  586594	      3114 ns/op	     913 B/op	       7 allocs/op
```

This is not a definitive solution to CPU usage increase, but it's a step further 😄

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Library%2Fgo-libaudit-mbp%2FPR-111/detail/PR-111/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Library%2Fgo-libaudit-mbp%2FPR-111/detail/PR-111/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Library%2Fgo-libaudit-mbp%2FPR-111/detail/PR-111/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Library%2Fgo-libaudit-mbp%2FPR-111/detail/PR-111/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://go-libaudit_111.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-06-06T16:14:15.908Z&rangeTo=2022-06-06T16:34:15.908Z&transactionName=Library/go-libaudit-mbp/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=d8ba05179f36384c194c819ea1d46b40&transactionId=f482e1c46f2f9d06)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-06-06T16:24:15.908+0000


* Duration: 2 min 28 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 386  |
| Skipped      | 40 |
| Total        | 426   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->













<!--COMMENT_GENERATED_WITH_ID_comment.id-->

andrewkroh commented:
Comparing main (48e317eaca4fa4dfde10a84d070a084aed1ddfa6) to this branch using the existing benchmarks:

`go test -bench . -benchtime=10s -benchmem -run NONE ./auparse ./aucoalesce/`

```
benchmark                             old ns/op     new ns/op     delta
BenchmarkParseAuditHeader-10          143           144           +0.56%
BenchmarkParseAuditHeaderRegex-10     540           543           +0.46%
BenchmarkParseLogLine-10              361           358           -0.69%
BenchmarkCoalesceMessages-10          3114          3128          +0.45%

benchmark                             old allocs     new allocs     delta
BenchmarkParseAuditHeader-10          3              3              +0.00%
BenchmarkParseAuditHeaderRegex-10     2              2              +0.00%
BenchmarkParseLogLine-10              4              4              +0.00%
BenchmarkCoalesceMessages-10          13             13             +0.00%

benchmark                             old bytes     new bytes     delta
BenchmarkParseAuditHeader-10          21            21            +0.00%
BenchmarkParseAuditHeaderRegex-10     128           128           +0.00%
BenchmarkParseLogLine-10              374           358           -4.28%
BenchmarkCoalesceMessages-10          2718          2718          +0.00%
```

andrewkroh commented on auparse/auparse_test.go:
```diff
@@ -123,41 +142,182 @@ func TestExtractKeyValuePairs(t *testing.T) {
+				b.StartTimer()
+				out := extractKeyValuePairs(tc.in)
+				b.StopTimer()
+				if !assert.Equal(b, tc.out, out, "failed on: %v", tc.in) {
```
I think it is best to have a separate unit test that checks the output. This way we can keep the loop closely focused on the code being benchmarked and you don't have to use StartTime / StopTimer. Using `StartTime` and `StopTimer` within the loop is not ideal because it breaks the estimated number of iterations (since the  StartTime/StopTimer take a long time relative to extractKeyValuePairs).

When running `go test -run NONE -bench ExtractKeyValuePairs/bpf_message_current -benchtime=10s -benchmem ./auparse/` it takes close to two minutes. With the start/stop and assert removed the same benchmark task ~11s and completes a similar number of iterations (~1.1M).

efd6 replied:
If there are no side-effects the compiler is allowed to elide the code run in the benchmark. This is unlikely to happen here soon, but it is a possibility. A simple conditional will prevent this, something like `if len(out) != len(tc.out) { b.Errorf("unexpected result length: got:%d want:%d", len(out), len(tc.out) }` would do the trick.

Same below.

inge4pres replied:
I removed the assertion altogether, because re-reading I found that `TestExtractKeyValuePairs` is already covering that.

andrewkroh reviewed (commented):
Thanks for looking at the number of allocation. I haven't reviewed the actually changes yet. I was first measuring the impact. Leaving some minor comments on the benchmarks first.

efd6 commented on auparse/auparse.go:
```diff
@@ -66,10 +65,31 @@ type field struct {
+
+func (fm fieldMap) setFieldValue(key, value string) {
+	if f, ok := fm[key]; ok {
+		fm[key] = field{f.orig, value}
```
This *is* a local type, but the fields are the same type, so please add explicit field names.

efd6 commented on auparse/auparse.go:
```diff
@@ -66,10 +65,31 @@ type field struct {
+	return
+}
+
+func (fm fieldMap) find(key string) (field, error) {
+	if f, ok := fm[key]; ok {
+		return f, nil
+	}
+	return field{}, fmt.Errorf("%s key not found", key)
+}
```
The only error here is absence, so an `ok bool` would suffice. The absence errors can then be statically defined errors returned by the callers in most cases.

efd6 replied:
```
func (fm fieldMap) find(key string) (field, bool) {
	 f, ok := fm[key]
 	 return f, ok
}
```

inge4pres replied:
Apologies, but I don't think I understand your concern in keeping the method as initially proposed in the PR. 
Is it related to performance / readability / or?
The compiler says that only `key` escapes to the heap, and the error stays on the stack.
IIRC I saw the errors also escaping on the heap with the current implementation, but I may be wrong.

The reason for `find` to exist is to check the existence of a key and avoiding repeating the error formatting in every call site.
If `find` would only return the result of map fetch `fm[key]`, it would means that:

* all the callers have to check if `ok` is true _and_
* possibly return an error, that only has the `key` string as difference

This method's signature includes the error to limit the duplication in all the callers at the bottom of the file, IMO they are more readable now, compared with the current version, because the error checking is extracted in `find`.

efd6 replied:
OK. It's a trade-off (though note that the call to `fmt.Errorf` [imposes other costs](https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/fmt/errors.go;l=17-30)); though in the case that the errors are made at the caller site, there is no formatting since the string is already known, and so can be constructed during init, once. I'm not particularly fussed by this since I would expect the error path to be infrequent, but if it weren't, I would prefer the cost to be in the source and not in the runtime.

efd6 replied:
As I say, I'm not that fussed. The way that I had intended it would have had the errors constructed at init and stored as vars, so their allocation would have been amortised to O(0). The code as it is, requires allocation of a `fmt.pp` and parsing of a format string (some of the additional complexity can be seen [here](https://godbolt.org/z/x5Eq43Y6d), but godbolt doesn't show the work that is done in fmt), however the error path is expected to be infrequent, so it's not a huge issue. If it were more likely, I would push back harder.

efd6 commented on auparse/auparse.go:
```diff
@@ -514,78 +533,78 @@ func execveArgs(data map[string]*field) error {
+	if err != nil {
+		field, err = fm.find("res")
+		if err != nil {
 			return errors.New("success and res key not found")
```
Make this a package var.

inge4pres replied:
I didn't change this specific error in this PR, only the logic around it, but yeah it makes sense to add a var

efd6 commented on auparse/auparse_test.go:
```diff
@@ -204,55 +364,60 @@ func NewStoredAuditMessage(msg *AuditMessage) *StoredAuditMessage {
+	var messages []*AuditMessage
+	s := bufio.NewScanner(bufio.NewReader(f))
+	var lineNum int
+	for s.Scan() {
```
Check `sc.Err()` before return.

efd6 commented on auparse/auparse_test.go:
```diff
@@ -123,41 +142,182 @@ func TestExtractKeyValuePairs(t *testing.T) {
+	}
+
+	for _, tc := range tests {
+		tc := tc
```
These copies are not needed here and below; the closure completes before the next iteration.

inge4pres commented on auparse/auparse_test.go:
```diff
@@ -123,41 +142,182 @@ func TestExtractKeyValuePairs(t *testing.T) {
+
+	for _, tc := range tests {
+		tc := tc
+		b.Run(tc.name+"_current", func(b *testing.B) {
```
Ah good spot 👍🏼 it's a leftover from previous tests, will amend in latest commit

inge4pres commented:
Re-requesting review, I think I addressed all the comments in https://github.com/elastic/go-libaudit/pull/111/commits/3692eb071a55567ecd91a50215f4d542ee4d1f6a, let me know otherwise 🙏🏼

efd6 commented on auparse/auparse_test.go:
```diff
@@ -123,41 +142,176 @@ func TestExtractKeyValuePairs(t *testing.T) {
+	tests := dataTests
+	for _, tc := range tests {
+		tc := tc
+		b.Run(tc.name, func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				b.StopTimer()
+				var (
+					data map[string]string
+					err  error
+				)
+				msg, errP := ParseLogLine(tc.text)
+				if errP != nil {
+					b.Fatalf("parsing message: %v", errP)
+				}
+				b.StartTimer()
+				data, err = msg.Data()
+				b.StopTimer()
+				require.Nil(b, err)
+				assert.Equal(b, tc.output, data)
+			}
+		})
```
The `ParseLogLine` call can be hoisted out of the loop and the check probably should be in a test rather than in the benchmark, just leaving a nil check in the benchmark to avoid the compiler optimising it away.

```suggestion
		msg, err := ParseLogLine(tc.text)
		if err != nil {
			b.Errorf("parsing message: %v", err)
			continue
		}
		b.Run(tc.name, func(b *testing.B) {
			for i := 0; i < b.N; i++ {
				_, err = msg.Data()
				if err != nil {
					b.Fatalf("unexpected error: %v", err)
				}
			}
		})
```

inge4pres replied:
Thanks for the suggestion 👍🏼 looks nicer

inge4pres replied:
Actually I noticed this output after applying your suggestion, it does not seem correct

```
BenchmarkAuditMessageData/syscall-8                 	1000000000	         2.342 ns/op	       0 B/op	       0 allocs/op
BenchmarkAuditMessageData/user_cmd-8                	1000000000	         2.352 ns/op	       0 B/op	       0 allocs/op
```

Not sure what's happening: could it be that we don't make any use of the map returned by `Data()`, and that makes it not part of the calculation of the benchmark time/memory?

inge4pres replied:
I had to revert your suggestion, the previous form is the only one that allows the benchmark tests to run and return meaningful results.

efd6 replied:
> Not sure what's happening: could it be that we don't make any use of the map returned by Data(), and that makes it not part of the calculation of the benchmark time/memory?

The SSA passes won't omit part of the return, examining the `err` should be enough to ensure that the code is run. What happens if the err is declared in the loop? What happens if you only hoist out of the loop and not out of the closure?

efd6 commented on auparse/auparse_test.go:
```diff
@@ -123,41 +142,176 @@ func TestExtractKeyValuePairs(t *testing.T) {
+				t.Fatalf("parsing message: %v", err)
+			}
+			data, err := msg.Data()
+			require.Nil(t, err)
+			assert.Equal(t, tc.output, data)
```
Check in a test rather than in a benchmark.

efd6 replied:
Sorry, too early. Ignore this.

efd6 commented on auparse/auparse.go:
```diff
@@ -290,222 +308,224 @@ func normalizeAuditMessage(typ AuditMessageType, msg string) (string, error) {
 
 // Enrichment after KV parsing
-
-//nolint:errcheck // Continue enriching even if some fields do not exist.
```
I prefer the linter comment to blank identifier assignment; it makes it clear why the errors are being ignored.

inge4pres replied:
Do you mean keeping 
```
// Continue enriching even if some fields do not exist.
```
as a separate comment line?
The lint exclusion `//nolint:errcheck` is not necessary anymore.

inge4pres commented on auparse/auparse_test.go:
```diff
@@ -123,41 +142,176 @@ func TestExtractKeyValuePairs(t *testing.T) {
+func BenchmarkAuditMessageData(b *testing.B) {
+	tests := dataTests
+	for _, tc := range tests {
+		tc := tc
```
Ah sorry, missed that 🙏🏼

efd6 reviewed (approved):
LGTM, but please wait for @andrewkroh.

andrewkroh reviewed (approved):
Can you please add an entry to the CHANGELOG.md in the root of the repo.

> do you usually squash PR commits? 

Yeah, we normally let GH squash it.

inge4pres commented:
> Can you please add an entry to the CHANGELOG.md in the root of the repo.
> 
> > do you usually squash PR commits?
> 
> Yeah, we normally let GH squash it.

CHANGELOG line added in https://github.com/elastic/go-libaudit/pull/111/commits/8c93c65c9230447d747bfd2c7942964f85cb5cb4.
Thanks both for your help 🙏🏼
